### ISSUE TITLE
iOS app signed with invalid signature since ios 16 release and xcode 14

### ISSUE THREAD

**Issue by tpitman** [xamarin/xamarin-macios#16190]
### Steps to Reproduce

1. Set Apple signing to automatic with correct team for app bundle id
2. Build and run to actual ios device

### Expected Behavior
App built and signed and debugging on ios device.

### Actual Behavior
App builds, but during deploy the following error is output in the deploy window and the app is not installed:

ApplicationVerificationFailed: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.LjPGBm/extracted/DynoJet.iOS.app : 0xe8008014 (The executable contains an invalid signature.)
error MT1006: Could not install the application '[my app path and name here]' on the device '[device name here]': AMDeviceSecureInstallApplicationBundle returned: 0xe8008014.

### Environment

I had installed xcode 14, but that didn't even build, so I installed and am using xcode 13.4.1 as recommended.

<details>
<summary>Version information</summary>

```
Visual Studio Enterprise 2022 for Mac Preview
Version 17.4 Preview (17.4 build 1583)
Installation UUID: 90f578df-1f3b-4f3e-9726-e7054797f71f

Runtime
.NET 6.0.5 (64-bit)
Architecture: Arm64

Roslyn (Language Service)
4.4.0-2.22419.1+ae5663806be933515723b2d5a85faf5d378f460d

NuGet
Version: 6.3.0.128

.NET SDK (Arm64)
SDK: /usr/local/share/dotnet/sdk/7.0.100-rc.1.22431.12/Sdks
SDK Versions:
	7.0.100-rc.1.22431.12
	6.0.401
	6.0.400
	6.0.400-preview.22330.6
MSBuild SDKs: /Applications/Visual Studio (Preview).app/Contents/MonoBundle/MSBuild/Current/bin/Sdks

.NET SDK (x64)
SDK Versions:
	6.0.401
	6.0.400
	6.0.400-preview.22330.6
	6.0.301
	6.0.300
	6.0.203
	6.0.202
	6.0.104
	6.0.103
	6.0.102
	6.0.101
	5.0.408
	5.0.407
	5.0.406
	5.0.405
	5.0.404
	3.1.423
	3.1.422
	3.1.421
	3.1.420
	3.1.419
	3.1.418
	3.1.417
	3.1.416

.NET Runtime (Arm64)
Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	7.0.0-rc.1.22426.10
	6.0.9
	6.0.8
	6.0.6
	6.0.5
	6.0.4
	6.0.1

.NET Runtime (x64)
Runtime: /usr/local/share/dotnet/x64/dotnet
Runtime Versions:
	6.0.9
	6.0.8
	6.0.6
	6.0.5
	6.0.4
	6.0.3
	6.0.2
	6.0.1
	5.0.17
	5.0.16
	5.0.15
	5.0.14
	5.0.13
	3.1.29
	3.1.28
	3.1.27
	3.1.26
	3.1.25
	3.1.24
	3.1.23
	3.1.22

Xamarin.Profiler
Version: 1.8.0.19
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

Updater
Version: 11

Apple Developer Tools
Xcode 13.4.1 (20504)
Build 13F100

Xamarin.Mac
Version: 8.12.0.2 (Visual Studio Enterprise)
Hash: 87f98a75e
Branch: d17-3
Build date: 2022-07-25 20:18:54-0400

Xamarin.iOS
Version: 15.12.0.2 (Visual Studio Enterprise)
Hash: 87f98a75e
Branch: d17-3
Build date: 2022-07-25 20:18:55-0400

Xamarin Designer
Version: 17.4.0.124
Hash: ffbba0c031
Branch: remotes/origin/d17-4
Build date: 2022-09-15 15:36:42 UTC

Xamarin.Android
Version: 13.0.99.81 (Visual Studio Enterprise)
Commit: xamarin-android/main/7c9c24b
Android SDK: /Users/tpitman/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		12.0 (API level 31)
		8.1  (API level 27)
		11.0 (API level 30)
		10.0 (API level 29)
		9.0  (API level 28)
		13.0 (API level 33)
		7.0  (API level 24)

SDK Command-line Tools Version: 7.0
SDK Platform Tools Version: 33.0.2
SDK Build Tools Version: 33.0.0

Build Information: 
Mono: dffa5ab
Java.Interop: xamarin/java.interop/main@2c8b0a8f
SQLite: xamarin/sqlite/3.39.2@40e8743
Xamarin.Android Tools: xamarin/xamarin-android-tools/main@29f11f2

Microsoft Build of OpenJDK
Java SDK: /Library/Java/JavaVirtualMachines/microsoft-11.jdk
11.0.12
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Eclipse Temurin JDK
Java SDK: /Library/Java/JavaVirtualMachines/temurin-8.jdk
1.8.0.302
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Android SDK Manager
Version: 17.4.0.45
Hash: 8cac7ea
Branch: remotes/origin/HEAD~1
Build date: 2022-09-15 15:36:46 UTC

Android Device Manager
Version: 0.0.0.1196
Hash: b2faf27
Branch: main~1
Build date: 2022-09-15 15:36:46 UTC

Build Information
Release ID: 1704001583
Git revision: 70322329919c72a11d9795a3cb403fe4115a5878
Build date: 2022-09-15 15:34:40+00
Build branch: release-17.4
Build lane: release-17.4

Operating System
Mac OS X 12.6.0
Darwin 21.6.0 Darwin Kernel Version 21.6.0
    Mon Aug 22 20:19:52 PDT 2022
    root:xnu-8020.140.49~2/RELEASE_ARM64_T6000 arm64
```

</details>

### Build Logs

[iOSBuildLog_20220928.txt](https://github.com/xamarin/xamarin-macios/files/9667438/iOSBuildLog_20220928.txt)

**Comment by dalexsoto** [comment]
Does your device have iOS 16? Looks like Apple now requires having Developer Mode enabled https://developer.apple.com/documentation/xcode/enabling-developer-mode-on-a-device

**Comment by tpitman** [comment]
My device is running iOS version 15.6.1, so it should not need developer mode.

**Comment by dalexsoto** [comment]
do you have the "Apple Developer Account" setup in your VS settings to auto provision? (wether is on VS Windows or VS Mac)?

![image](https://user-images.githubusercontent.com/204671/192929582-e3d7784f-ca62-459e-b807-617e128b7aba.png)

**Comment by tpitman** [comment]
Yes I do. I don't think I would be able to choose Automatic Provisioning and have teams in the drop down list and be able to pick one and have it succeed in doing the provisioning if I didn't have that set up.

**Comment by decv86** [comment]
+1 here, Xcode 14 simple project work fine, but VS get this error. Apple account have + download all profiles iOS, but error save... need help!!!

**Comment by Kalyxt** [comment]
I have the same problem, but only with one project which is strange at least.
What I did in order to fix it.
- Deleted ALL development certs on developer.apple
- Deleted ALL provisioning profiles with my APP ID
- Deleted %LOCALAPPDATA%\Xamarin\iOS\Provisioning
- Log in as Apple developer in VS settings and connected to Mac
- Opened VS 2022 and selected Automatic Provisioning, it created all things needed to debug
Result:
ApplicationVerificationFailed: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.vllEGR/extracted/APPNAME.iOS.app : 0xe8008014 (The executable contains an invalid signature.)

Lets try luck with VS 2022 for Mac preview
Everything seems right
Signing identity - Developer: Name
Provisioning profile - VS: domain.company.appname Development

and what I get when I try to debug ? 
Please Start Application

Visual Studio cannot start the application automatically because it was signed with a Distribution profile. Please start it tapping the application icon on the device.

HOW?? I don't even have Distribution cert and profile!
When I look in Deploying to device output - 
ApplicationVerificationFailed: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.IYeNQS/extracted/APPNAME.iOS.app : 0xe8008014 (The executable contains an invalid signature.)
error MT1006: Could not install the application '/Users/username/Projects/Xamarin/APPNAME/APPNAME.iOS/bin/iPhone/Debug/device-builds/ipad13.16-15.7/APPNAME.iOS.app' on the device 'User iPad M1': AMDeviceSecureInstallApplicationBundle returned: 0xe8008014.

**Comment by decv86** [comment]
> I have the same problem, but only with one project which is strange at least. What I did in order to fix it.
> 
> * Deleted ALL development certs on developer.apple
> * Deleted ALL provisioning profiles with my APP ID
> * Deleted %LOCALAPPDATA%\Xamarin\iOS\Provisioning
> * Log in as Apple developer in VS settings and connected to Mac
> * Opened VS 2022 and selected Automatic Provisioning, it created all things needed to debug
> * Open XCode and download all profiles
>   Result:
>   ApplicationVerificationFailed: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.vllEGR/extracted/APPNAME.iOS.app : 0xe8008014 (The executable contains an invalid signature.)
> 
> Lets try luck with VS 2022 for Mac preview Everything seems right Signing identity - Developer: Name Provisioning profile - VS: domain.company.appname Development
> 
> and what I get when I try to debug ? Please Start Application
> 
> Visual Studio cannot start the application automatically because it was signed with a Distribution profile. Please start it tapping the application icon on the device.
> 
> HOW?? I don't even have Distribution cert and profile! Deploying to device output - ApplicationVerificationFailed: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.IYeNQS/extracted/APPNAME.iOS.app : 0xe8008014 (The executable contains an invalid signature.) error MT1006: Could not install the application '/Users/username/Projects/Xamarin/APPNAME/APPNAME.iOS/bin/iPhone/Debug/device-builds/ipad13.16-15.7/APPNAME.iOS.app' on the device 'User iPad M1': AMDeviceSecureInstallApplicationBundle returned: 0xe8008014.

I did the same - but there is no result ... I know that it works with iOS version: 15.4.303 - but this is not a solution, I need to test on iOS 16

**Comment by dalexsoto** [comment]
> I did the same - but there is no result ... I know that it works with iOS version: 15.4.303 - but this is not a solution, I need to test on iOS 16

@decv86 Your issue seems different, can you try this in your iOS 16 device? https://developer.apple.com/documentation/xcode/enabling-developer-mode-on-a-device

**Comment by decv86** [comment]
> > I did the same - but there is no result ... I know that it works with iOS version: 15.4.303 - but this is not a solution, I need to test on iOS 16
> 
> @decv86 Your issue seems different, can you try this in your iOS 16 device? https://developer.apple.com/documentation/xcode/enabling-developer-mode-on-a-device

I testing iPhone SE, iPhone 7 = iOS 15 and iPhone 13 mini = iOS 16. Yes - i enable developer mode. And how i say, xcode 14 deploy test project with not problem.

**Comment by decv86** [comment]
if you need logs or any information, feel free to tell me, I can provide remote access, I spent many hours solving the problem ...

**Comment by rolfbjarne** [comment]
It's getting complicated figuring out who said what and who answered which question, so if you're not the original reporter (@decv86, @Kalyxt) , please file a new issue with your info and we'll help you there.

@tpitman:

* Does it happen with just one project or any project?. i.e. can you launch a new app from a template?
* Can you create a new project in Xcode and launch that on device?

**Comment by decv86** [comment]
I'm sure I'm facing the same problem as the author of the post

**Comment by tpitman** [comment]
@rolfbjarne It seems to be all of my apps that won't sign and deploy.

I am able to build, sign and deploy my xcode projects that are not done with VS.

**Comment by rolfbjarne** [comment]
@tpitman do you know when this started to happen? Did you update VSMac? Or was it something else?

**Comment by tpitman** [comment]
About the time that Apple released iOS 16 and their new dev tools. I also updated VSMac when the last 2.1 preview update came out. I also updated the MacOS with its latest patch for 12.6 and there was a command line tools update at the same time.

I downloaded the xcode 14 and install it, but then downloaded the 13.4.1 version because VS has trouble compiling with xcode 14.

**Comment by rolfbjarne** [comment]
Can you check your project files and verify that you don't have the `EnableCodeSigning` property set to `false`? If you do, please change it to `true`:

```xml
<PropertyGroup>
    <EnableCodeSigning>true</EnableCodeSigning>
</PropertyGroup>
```

and then please let us know if that was the problem.

**Comment by tpitman** [comment]
@ArcanoxDragon What property group did you find that in your project file?

**Comment by ArcanoxDragon** [comment]
> @ArcanoxDragon What property group did you find that in your project file?

It exists once in every configuration-specific `PropertyGroup` automatically generated by Studio (e.g. `<PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'DEV|iPhone'">` for our `DEV` configuration on the `iPhone` platform).

**Comment by tpitman** [comment]
My project file doesn't contain it anywhere. Very interesting.

**Comment by tj-devel709** [comment]
@tpitman Could you try adding just 
```
<PropertyGroup>
    <EnableCodeSigning>true</EnableCodeSigning>
</PropertyGroup>
```
to its own property group and let us know the results? Just be sure that "EnableCodeSigning" is not being set to false anywhere else

**Comment by tpitman** [comment]
That fixed it!

So it looks like the main issue is that property was somehow not required before where now it is.

Keep in mind I used to be able to debug this project on iOS without issue just a couple weeks ago. It was only after one of the latest updates to 17.4 preview that it broke. My project file has not changed, so having that property must not have been required before to sign and debug.

**Comment by znelson32** [comment]
> That fixed it!
> 
> So it looks like the main issue is that property was somehow not required before where now it is.
> 
> Keep in mind I used to be able to debug this project on iOS without issue just a couple weeks ago. It was only after one of the latest updates to 17.4 preview that it broke. My project file has not changed, so having that property must not have been required before to sign and debug.

Mine were set to 'false' forever as well. Changing it to 'true' gets me past the signing error on my app. I had some similar complaints with third party assemblies but updating those to latest-stable resolved my issues.

**Comment by Kalyxt** [comment]
Setting this property to `true` fixed this issue for me too. I had it like this for some reason

```
<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|iPhone' ">
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\iPhone\Debug</OutputPath>
    <DefineConstants>DEBUG</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
    <MtouchArch>ARM64</MtouchArch>
    <CodesignKey>Apple Development:</CodesignKey>
    <MtouchDebug>true</MtouchDebug>
    <CodesignEntitlements>Entitlements.plist</CodesignEntitlements>
    <CodesignProvision>VS:  Development</CodesignProvision>
    <MtouchSdkVersion>15.0</MtouchSdkVersion>
    <EnableCodeSigning>false</EnableCodeSigning>
  </PropertyGroup>
```

Thanks!

**Comment by rolfbjarne** [comment]
@tpitman 

> That fixed it!
> 
> So it looks like the main issue is that property was somehow not required before where now it is.
> 
> Keep in mind I used to be able to debug this project on iOS without issue just a couple weeks ago. It was only after one of the latest updates to 17.4 preview that it broke. My project file has not changed, so having that property must not have been required before to sign and debug.

It's really puzzling that you didn't have the property, and that adding it fixed the problem, for a few reasons:

1. You're using a build of Xamarin.iOS that's over 2 months old (the commit is from July 26th):
    > Xamarin.iOS 15.12.0.2 (d17-3: 87f98a75e)
    It seems odd that it suddenly started failing (even if you just installed this particular update, other people should have run into problems earlier if the default really changed from `true` to `false`).
2. There's a known bug with VSMac, where VSMac writes out the property if it's not in the csproj (and sets it to false). This seems to be what's happening with everybody else.
3. As far as I can tell the default is (still) `true` when building for device, and I'm unable to reproduce what you're seeing.

Would it be possible for you to build from the command line with the property removed from your project file, [get a binary build log](https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#binary-build-logs) and upload that?

Something like this:

```shell
# clean the project
$ git clean -xfd # (or whatever method works for you)

# build the project for device
$ msbuild myproject.csproj /p:Platform=iPhone /bl:msbuild.binlog
[build outout]

# try to install the app on the device to confirm it's broken
$ /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/bin/mlaunch --installdev bin/iPhone/Debug/*.app
[this should fail]

# the binary build log to upload is "msbuild.binlog"
```

**Comment by tpitman** [comment]
I don't know what happened specifically, but after completing some work I pushed some bug fixes to our git repository and went back to the develop branch which doesn't have this fix in it.

I read your post and was going to follow it.

I opened the project file to edit and the setting was gone as expected.

I removed the app from my iPhone and did a build.

This time it built and deployed without issue and without having that setting.

I have no idea what could have changed.

Sorry I can't help figure that out as it is all working now without that property.

**Comment by rolfbjarne** [comment]
> I don't know what happened specifically, but after completing some work I pushed some bug fixes to our git repository and went back to the develop branch which doesn't have this fix in it.
> 
> I read your post and was going to follow it.
> 
> I opened the project file to edit and the setting was gone as expected.
> 
> I removed the app from my iPhone and did a build.
> 
> This time it built and deployed without issue and without having that setting.
> 
> I have no idea what could have changed.

This is why programming is fun 😁 

Anyways, I'm going to close this, since we already have another bug open for VSMac saving the wrong value (https://developercommunity.visualstudio.com/t/Code-signing-disabled-for-new-projects/10163187).

Feel free to comment here or open a new issue if you ever figure out what happened.
